[feature fix] Enable markup-safe rendering as flag in route declaration (markup-1)

abought opened the pull request:
## Purpose
Past efforts to enable markup-safe escaping by default have been sidetracked by the scope of changes needed. (#1308, @lyndsysimon )

This PR modifies the existing `OsfWebRenderer` class so that we can incrementally move views over to markup-safe rendering in a controlled fashion (one at a time) without swamping QA in a single global change.

This PR can be merged separate from any individual page conversion. The flag is disabled by default, so existing code should not be affected.

## Summary of changes
The `OsfWebRenderer` class can now be instantiated with an optional `safe=True` argument in `website/routes.py`:

`OsfWebRenderer('profile.mako', safe=True)`

This will cause all strings in that template (and sub-templates) to be rendered with the markup-safe "`h`" filter turned on by default. It can be explicitly disabled for any individual variable by using the "`n`" filter.

An optional `safe=True` argument (false by default) was also added to status messages used by `alert.mako`. Messages that require this (based on #1308) can be converted as needed to avoid display artifacts.

Aside from these changes, a new "sjson" filter has been implemented and made available to templates. This is slightly safer against markup than normal `json.dumps`, and would simplify the code in a variety of templates. 

## Further notes
- Tracking issue for overall task: #2933 
- Previous effort: #1308
- http://docs.makotemplates.org/en/latest/filtering.html
- DB fix script https://github.com/CenterForOpenScience/osf.io/pull/1308/files#diff-04f425db43b165f8c694fd21eaca0ed8

abought commented:
I've updated this PR to resolve merge conflicts.

This PR is basically groundwork- as far as I can tell it wouldn't change any existing pages unless the `safe=True` flag was added to that route. But it would give people access to a more consistent JSON filter (deprecating js_str and json.dumps).

What would need to be done to see this PR move forward?

lbanner commented:
@sloria ... Let's get this PR of @abought 's reviewed as a priority PR -- There is good security value here, and would like @abought not to have to keep bringing this up-to-date if we can get it reviewed + deployed. Thanks.

sloria commented:
Marked for 0.41.0

lyndsysimon commented on framework/routing/__init__.py:
```diff
@@ -187,15 +197,25 @@ def render_jinja_string(tpl, data):
-def render_mako_string(tpldir, tplname, data):
+def render_mako_string(tpldir, tplname, data, safe=False):
+    """Render a mako template. Optional safe argument to activate markup escaping"""
+    if safe is True:
```
Could this conditional be replaced with `lookup_obj = _tpl_lookup_safe if safe else _tpl_lookup`?

I don't see how `default_filters` is useful later. Passing `lookup_obj` to the `Template` instantiation should set the filters. If it doesn't, then they can be accessed directly at `lookup_obj['template_args']['default_filters']`.

abought replied:
It's more concise and avoids duplication; I like it. Changes pushed.

lyndsysimon commented on framework/routing/__init__.py:
```diff
@@ -392,12 +413,15 @@ def __init__(self, template_name, renderer=None, error_renderer=None,
                      to add to data from view function
         :param detect_render_nested: Auto-detect renderers for nested
             templates?
+        :param safe: Boolean: turn on markup-safe escaping?
```
In Mako, the "html safe" filter is applied when the data is known to be safe, not when you want to make it safe. I think using the word "safe" to mean the opposite here could be confusing. Perhaps flipping the boolean and renaming it "unsafe" would work?

abought replied:
Interesting comment on ambiguous meaning; in this context I had intended it to mean "apply markupsafe filters". I think this meaning may be closer to how the Mako documentation uses the word safe: http://docs.makotemplates.org/en/latest/filtering.html

`h : HTML escaping, provided by markupsafe.escape(string)`

As opposed to turning it off, which is 
`n : disable all default filtering; only filters specified in the local expression tag will be applied.`

Though the Jinja2 documentation uses the word safe in an opposite manner, as you note. I'm not sure what a neutral term would be...?

lyndsysimon replied:
Now that `Status` uses the parameter `trust`, is there any reason not to go with that name here as well?

abought replied:
That's a good question. The names were intentionally made different, because they represent different and opposite behaviors:

- `safe` refers to turning on markupsafe escaping. We want it to be True by default. (so that people who override it are **literally** saying "safe=False")
- `trust` refers to turning *off* markupsafe escaping. We want it to be False by default (people must explicitly indicate that the content can be trusted as raw HTML)

Because the status tuple is just data (not HTML), we judged that the word "safe" was not as meaningful in that different context. Hence the different semantics and defaults. I'm probably overthinking this.

lyndsysimon commented:
My one comment aside, this looks great. It's definitely an improvement over what we have now.

abought commented:
> My one comment aside, this looks great. It's definitely an improvement over what we have now.

Just to manage expectations... remember that this PR doesn't actually make *anything* markup safe by itself. It provides an option (False by default) so that we can start migrating separate views over in the future, as well as a new `sjson` filter for use in templates. This is infrastructure, not a change by itself.

Once all of the OSF pages are migrated incrementally, we can change the default to True. Meanwhile it will let us make new content that's designed to be safe from the start.

sloria commented on website/templates/alert.mako:
```diff
@@ -1,12 +1,17 @@
         % endif
-        <p>${message}</p>
+
+        % if safe:
```
Discussed in person, but the semantics of the `safe` parameter is different from the `safe` parameter for the template renderer. Maybe rename to `trust` or `is_html`.

sloria commented on framework/routing/__init__.py:
```diff
@@ -30,6 +30,16 @@
     module_directory='/tmp/mako_modules'
 )
+
+_tpl_lookup_safe = TemplateLookup(
```
As discussed in person, this might not be necessary. Worth looking into whether we need this

abought replied:
As far as I can tell the separate objects are needed right now: `TemplateLookup` creates `Template`s directly using its own `default_filters` args, rather than inheriting from the parent.

lyndsysimon commented on website/util/sanitize.py:
```diff
@@ -94,3 +97,16 @@ def safe_unescape_html(value):
     return value
+
+
+def safe_json(value):
```
I'd like to change this to `to_json`, `as_json`, or something similar - both here and in `get_globals` in `routes.py`.

Of the two changes, I feel most strongly about the naming inside of a Mako template context. The function itself can have a more explicit name if that makes more sense.

lyndsysimon commented on framework/routing/__init__.py:
```diff
@@ -187,15 +198,20 @@ def render_jinja_string(tpl, data):
 
     tpl = mako_cache.get(tplname)
     if tpl is None:
+        with open(os.path.join(tpldir, tplname)) as f:
```
Oooh - good catch.

This sort of thing I believe is the root of the "Too many open files" errors many of us have been getting while running the OSF.

lyndsysimon commented:
Good to go!

abought commented:
Updated PR with proposed changes. In this round, added changes to `base.mako` to reflect proposed best practices for JS string handling.

I was starting to do this in the child PRs, but had never gone back and edited `base.mako`. Basically:

* We should not be building strings inside javascript. `jsvar = '${pythonvar}'` is vulnerable to people with apostrophes in their names.
* Strings inside javascript should never be HTML escaped, since they are not being used in HTML. We'll let JS handle its own escaping if it displays that text.
* Deprecate use of `js_str`, `json.dumps`, etc in templates in favor of `| sjson, n`

Also renamed some parameters. Impact: when QA looks at this, make sure that analytics etc are behaving normally; applying `sjson` isn't expected to break anything.

sloria commented:
Looks good. :octopus:
